scheduler/volumebinding: passive assume cache

huww98 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
/kind flake

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Currently volume and dynamic-resource plugin shares an AssumeCache implementation. However, they features significantly different use case.  DRA call Assume() on objects returned by APIServer, but volume call Assume() on objects yet to be sent to the APIServer.

VolumeBinding plugin only makes one update request, while DynamicResource makes 2 requests (add finalizer then update allocation status).  Taking advantage of this, currently the volume cache is simpler:

1. Reserve: assume PV/PVC will be updated
2. PreBind: really send the update request
3. AssumeCache receives an update from informer and overwrite the assumed state.
   - if PreBind succeeded, this will surely include the update from step 2.
   - if PreBind is not finished yet, and this is an irrelevant update, it is safe to overwrite the assumed state, because our update in PreBind will surely fail with Conflict.

While for DynamicResource:

1. Reserve: add devices to inFlightAllocations
2. PreBind:
   1. send the 2 update requests
   2. add the returned object into AssumeCache
   3. AssumeCache dispatch events synchronously to update allocatedDevices
   4. remove devices from inFlightAllocations

DynamicResource needs some features from AssumeCache that is not necessary for VolumeBinding:
1. DynamicResource needs strictly ordered update events to update allocatedDevices, including those from Assume() and Restore()
2. DynamicResource needs to compare ResourceVersion to prevent the assumed state from being overwritten by older version from informer.  While this works, the [doc][1] says: "you must not compare resource versions for greater-than or less-than relationships".

Given so much difference, it can be beneficial to fork another simpler AssumeCache for VolumeBinding plugin. Because of no need to send events, the lite AssumeCache is a passive component. It only record the assumed version without copying all objects from informer into its local cache. When reading, we read from both informer and local cache.  So it will always be up-to-date with informer, no need to wait for event handler.

This resolves a race condition where AssumeCache and scheduler queue both receive events from informer. When a pod is scheduling due to PV update event, the PVCache may be not updated yet because it has not processed the relevant event.

The lite version still listens events from informer, but only for cleaning up its local cache to save memory.

[1]: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
Fixes #133872
Fixes #133611

#### Special notes for your reviewer:

Replaces #133924

To resolve the race, alternatively, we can build the AssumeCache in pkg/scheduler.New, listening for events from AssumeCache instead of informer. Then pass the AssumeCache all the way down from `frameworkOptions`, `frameworkImpl`, `Handle` interface to `VolumeBinding` plugin, like how DRA deal with this.

But I'd like to make VolumeBinding plugin more self-contained, and avoid introduce another dependencies on the framework.

Line of code:

`pkg/scheduler/framework/plugins/volumebinding/assume_cache_base.go`: 147
`pkg/scheduler/util/assumecache/assume_cache.go`: 342

The lite version is a lot simpler.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix occasional schedule delay when the static PV is created
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

pacoxu commented:
/test pull-kubernetes-unit
a known flake https://github.com/kubernetes/kubernetes/issues/130018

huww98 commented:
/retest
APIServer test failed, seems unrelated.

macsko commented:
/cc @sanposhiho

macsko commented on pkg/scheduler/framework/plugins/volumebinding/passive_assume_cache.go:
```diff
@@ -0,0 +1,221 @@
+			DeleteFunc: c.delete,
+		},
+	)
+	utilruntime.Must(err)
```
Please don't use `utilruntime.Must` in scheduler's code

macsko commented on pkg/scheduler/framework/plugins/volumebinding/passive_assume_cache.go:
Maybe move that to the `util/passiveassumecache`?

macsko commented on pkg/scheduler/framework/plugins/volumebinding/assume_cache.go:
```diff
@@ -20,112 +20,68 @@ import (
-	}
-	return pv, nil
-}
+func (c PVAssumeCache) ListPVs(storageClassName string) []*v1.PersistentVolume {
```
I'm not sure if `PVAssumeCache` should use  `passiveAssumeCache` internals. Maybe worth moving the `List` method to the cache?

macsko commented on pkg/scheduler/framework/plugins/volumebinding/assume_cache.go:
```diff
@@ -20,112 +20,68 @@ import (
+// NewPVAssumeCache creates a PV assume cache.
+func NewPVAssumeCache(logger klog.Logger, informer informer) PVAssumeCache {
+	logger = klog.LoggerWithName(logger, "pv-cache")
+	err := informer.GetIndexer().AddIndexers(map[string]cache.IndexFunc{
```
What's a benefit of replacing `cache.NewIndexer` to this (`AddIndexers`)?

macsko commented on pkg/scheduler/framework/plugins/volumebinding/assume_cache.go:
```diff
@@ -20,112 +20,68 @@ import (
 type PVAssumeCache struct {
-	*assumecache.AssumeCache
-	logger klog.Logger
+	*passiveAssumeCache[*v1.PersistentVolume]
```
What are the advantages of replacing the assume cache to the passive one? Now, we will need to support both and we need a strong reason to do that.

huww98 replied:
> Using a single event handler (like for DRA) would also avoid race condition.

Yes, as I wrote in the PR description, following DRA is also an option:
> To resolve the race, alternatively, we can build the AssumeCache in pkg/scheduler.New, listening for events from AssumeCache instead of informer. Then pass the AssumeCache all the way down from frameworkOptions, frameworkImpl, Handle interface to VolumeBinding plugin, like how DRA deal with this.

> But we would have to maintain two versions of caches - one for volume binding and one for DRA. Do we really have that strong reasons that would justify this approach?

I think we can discuss more on this. PassiveAssumeCache vs. follow DRA

Pros:
* Simpler from VolumeBinding PoV, less asynchronous, easier to reason about.
* No need to change scheduler framework for a single plugin, less coupled.
* Given the use case are significantly different between volume and DRA, using one implementation may limit the future development of either plugin, even causing unintended breakage. 
Cons:
* More code to maintain overall (+150 LoC). Not very much IMO.

msau42 commented:
> And the volume stress tests do not cover static PV cases. On the other do we need to cover stress tests for static PVs for this change?

For volume scheduler changes, we have integration stress tests (which is where the original test failure flake came from)

jsafrane reviewed (commented):
I kind of like the new assume cache. If we really need it is another question though.
@msau42  what do you think?

huww98 committed:
scheduler/volumebinding: use subtest

huww98 committed:
scheduler/volumebinding: pass testing.T to helper

huww98 committed:
scheduler/volumebinding: introduce testInformer

huww98 committed:
scheduler/volumebinding: passive assume cache

Currently volume and dynamic-resource plugin shares an AssumeCache
implementation. However, they features significantly different use case.  DRA
call Assume() on objects returned by APIServer, but volume call Assume() on
objects yet to be sent to the APIServer.

VolumeBinding plugin only makes one update request, while DynamicResource makes
2 requests (add finalizer then update allocation status).  Taking advantage of
this, currently the volume cache is simpler:

1. Reserve: assume PV/PVC will be updated
2. PreBind: really send the update request
3. AssumeCache receives an update from informer and overwrite the assumed state.
   a. if Prebind succeeded, this will surely include the update from step 2.
   b. if PreBind is not finished yet, and this is an irrelevant update, it is safe to
      overwrite the assumed state, because our update in PreBind will surely fail with Conflict.

While for DynamicResource:

1. Reserve: add devices to inFlightAllocations
2. PreBind:
   a. send the 2 update requests
   b. add the returned object into AssumeCache
   c. AssumeCache dispatch events synchronously to update allocatedDevices
   d. remove devices from inFlightAllocations

DynamicResource needs some features from AssumeCache that is not necessary for VolumeBinding:
1. DynamicResource needs strictly ordered update events to update allocatedDevices,
   including those from Assume() and Restore()
2. DynamicResource needs to compare ResourceVersion to prevent the assumed state from being
   overwritten by older version from informer.  While this works, the doc[1] says:
   "you must not compare resource versions for greater-than or less-than relationships".

Given so much difference, it can be beneficial to fork another simpler
AssumeCache for VolumeBinding plugin. Because of no need to send events, the
lite AssumeCache is a passive component. It only record the assumed version
without copying all objects from informer into its local cache. When reading,
we read from both informer and local cache.  So it will always be up-to-date
with informer, no need to wait for event handler.

This resolves a race condition where AssumeCache and scheduler queue both
receive events from informer. When a pod is scheduling due to PV update event,
the PVCache may be not updated yet because it has not processed the relevant
event.

The passive version still listens events from informer, but only for cleaning up
its local cache to save memory.

[1]: https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions

huww98 committed:
scheduler/volumebinding: remove Get[API]{PV,PVC}

should be replaced by generic Get[APIObj]

huww98 committed:
scheduler/volumebinding: target AssumeCache UT generic passiveAssumeCache

And removing duplicate tests

huww98 committed:
scheduler/volumebinding: rename passive_assume_cache_test.go

huww98 committed:
scheduler/volumebinding: add test for PVAssumeCache

huww98 commented on pkg/scheduler/framework/plugins/volumebinding/assume_cache.go:
```diff
@@ -20,112 +20,63 @@ import (
-		return nil, &assumecache.WrongTypeError{TypeName: "v1.PersistentVolumeClaim", Object: obj}
-	}
-	return pvc, nil
+func NewPVCAssumeCache(logger klog.Logger, informer informer) (PVCAssumeCache, error) {
```
OK, for the unit test part, I moved most cases to `passive_assume_cache_test.go`. I do the file rename in a dedicated commit, so `git blame` is still able to track this. On GitHub, reviewers can select the individual commit for clearer diff.

pohly commented:
> +1 to separate VolumeBinding and DRA's assume cache because VolumeBinding and DRA solve fundamentally different problems.
> 
> VolumeBinding is about data persistence. Its main job is to create a stable, long-term link between a pod and a specific piece of storage. DynamicResource is about access to hardware. Its job is to allocate a share of a node's specialized, often stateless, hardware (like GPUs or FPGAs) to a pod for the duration of its execution.

The question isn't about how DRA and VolumeBinding handle different things. It is about what they need from an assume cache. Is that really different? How?

IMHO it's very similar:
- track state before it gets sent to the API server ("in flight" in DRA)
- track items sent to the API server (assume cache)
- receive events (assume cache in DRA, informer in VolumeBinding but needs to be something else there)

huww98 commented:
> Implementing a client-side cache is absolutely nothing for the faint of heart.

Although it is named as "cache", this component is critical for making correct volume binding decisions, by taking account of our decisions in previous scheduling cycles. It is an account book for volume binding.

> Think very carefully whether you really can do it so that it has no race conditions!

We should provide 2 guarantees to user with the cache:
* After the informer dispatch a event, we guarantee the read object reversion >= the object triggers that event.
* The assumed object will finally be cleaned up.

We achieve this by maintaining these invariants:
* The revision of object in informer cache >= that in local cache
* if the revision in informer cache > local cache, there is at least one event from informer not processed yet by us.

Now we basically protect everything under the mutex, I think this is not very hard to reason about.

> Compare also against [mutation cache](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/mutation_cache.go)

I think the mutation cache is not suitable for us, because it uses LRU and is best-effort, while we need a guaranteed cache for correct decisions.

However, mutation cache also reads from both informer and local cache for each get request, matching the design of the proposed cache to some degree.

> IMHO it's very similar:
> * track state before it gets sent to the API server ("in flight" in DRA)
> * track items sent to the API server (assume cache)
> * receive events (assume cache in DRA, informer in VolumeBinding but needs to be something else there)

Yes, they are similar. But VolumeBinding only send one request per volume when binding. Taking advantage of this, the proposed passive assume cache plays the role of "in flight" + assume cache in DRA, fulfill the 3 requirements with one component, with less code, and do not require scheduler framework changes.

> You want some reviewer for SIG api-machinery for this.

OK, maybe I should send a message on their slack channel to gain some attention.

huww98 commented:
> Doesn't that mean that the change disappears? How is this handled?

@pohly  That would be the expected disappear. I quote from my PR description:

> 3. AssumeCache receives an update from informer and overwrite the assumed state.
>    - if PreBind succeeded, this will surely include the update from step 2.
>    - if PreBind is not finished yet, and this is an irrelevant update, it is safe to overwrite the assumed state, because our update in PreBind will surely fail with Conflict.

Your example fits the second condition. `PreBind` will fail, and we will call `Restore` anyway.

Rajalakshmi-Girish commented:
Eagerly waiting for this to merge, can’t wait to see `k8s.io/kubernetes/test/integration/volumescheduling.volumescheduling` finally stop flaking
https://github.com/kubernetes/kubernetes/issues/133611
@huww98

Rajalakshmi-Girish commented:
/test pull-kubernetes-integration-ppc64le
